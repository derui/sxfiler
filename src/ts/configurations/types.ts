import { ObjectEnum } from "@/utils";

export type ItemKey<T> = readonly [string, string, string];
export type SectionKey = readonly [string, string];
export type CategoryKey = readonly [string];

export type Descriptor<T> = {
  readonly key: T;
  readonly displayName: string;
  readonly description: string;
};

export type CategoryDescriptor = Descriptor<CategoryKey>;

export type SectionDescriptor = Descriptor<SectionKey>;

export type ItemDescriptor = Descriptor<ItemKey<any>>;

export type SelectOption = {
  value: any;
  toString(): string;
  display: string;
};

export const ItemElementType = {
  TEXT: "TEXT",
  NUMBER: "NUMBER",
  BOOLEAN: "BOOLEAN",
  SELECT_ONE: "SELECT_ONE",
} as const;
export type ItemElementType = ObjectEnum<typeof ItemElementType>;

export type ItemKind<K extends ItemElementType> = {
  kind: K;
};

export type TextItemType = ItemKind<"TEXT"> & {
  defaultValue?: string;
};
export type NumberItemType = ItemKind<"NUMBER"> & {
  defaultValue?: number;
};
export type BooleanItemType = ItemKind<"BOOLEAN"> & {
  defaultValue: boolean;
};
export type SelectOneItemType = ItemKind<"SELECT_ONE"> & {
  options: SelectOption[];
  defaultValue: string;
};

export type ItemType = TextItemType | NumberItemType | BooleanItemType | SelectOneItemType;

export type Item = ItemDescriptor & {
  readonly type: ItemType;
};

export type Section = SectionDescriptor & {
  readonly items: Item[];
};

export type Category = CategoryDescriptor & {
  readonly sections: Section[];
};
