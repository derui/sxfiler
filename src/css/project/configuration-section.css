.configuration-section {
  font-size: 1rem;
  grid-template-rows: 2rem 1fr;
  background-color: rgb(var(--color-base03));
  box-shadow: var(--box-shadow);
}

.configuration-section__header {
  position: relative;
  font-size: 1.5em;
  line-height: 1.5em;
  padding: var(--space-large) var(--space-base);
  color: rgb(var(--color-base3));
  background-color: rgb(var(--color-base02));

  /* border */
  box-shadow: var(--box-shadow);
}

.configuration-section__cell-container {
  padding-top: var(--space-base);
  padding-bottom: var(--space-base);
  background-color: rgb(var(--color-base03));
}
