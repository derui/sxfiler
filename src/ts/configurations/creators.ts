import { Section, Category, Item, ItemType, ItemKey, SectionDescriptor, CategoryDescriptor } from "./types";

export const createCategory = function (
  args: Omit<CategoryDescriptor, "key"> & { key: readonly [string]; sections: Section[] }
): Category {
  const { key, ...rest } = args;
  return { key: { key: key }, ...rest };
};

export const createSection = function (
  args: Omit<SectionDescriptor, "key"> & { key: readonly [string, string]; items: Item[] }
): Section {
  const { key, ...rest } = args;
  return { key: { key: key }, ...rest };
};

export const toItemKey = function <T = any>(key: [string, string, string]): ItemKey<T> {
  return {
    type_: undefined,
    key,
  };
};

export const createItem = function (args: {
  key: ItemKey<any>;
  displayName: string;
  description: string;
  type: ItemType;
}): Item {
  const { key, ...rest } = args;
  return {
    key,
    displayName: rest.displayName,
    description: rest.description,
    type: rest.type,
  };
};
