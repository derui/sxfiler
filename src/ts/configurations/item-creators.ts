import { ItemElementType, ItemType, SelectOption } from "./types";

export const createText = (value?: string): ItemType => ({ kind: ItemElementType.TEXT, defaultValue: value });

export const createNumber = (value?: number): ItemType => ({ kind: ItemElementType.NUMBER, defaultValue: value });

export const createBoolean = (value: boolean): ItemType => ({ kind: ItemElementType.BOOLEAN, defaultValue: value });

export const createSelectOne = (options: SelectOption[], value: string): ItemType => ({
  kind: ItemElementType.SELECT_ONE,
  options,
  defaultValue: value,
});
