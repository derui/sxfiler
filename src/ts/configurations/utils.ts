import { ItemElementType, Item } from "./types";

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const isTypedText = function (arg: Item, value: any): value is string {
  return arg.type.kind === ItemElementType.TEXT && value !== undefined && value !== null;
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const isTypedNumber = function (arg: Item, value: any): value is number {
  return arg.type.kind === ItemElementType.NUMBER && value !== undefined && value !== null;
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const isTypedBoolean = function (arg: Item, value: any): value is boolean {
  return arg.type.kind === ItemElementType.BOOLEAN && value !== undefined && value !== null;
};
