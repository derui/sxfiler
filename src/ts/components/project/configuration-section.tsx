import { h, FunctionalComponent } from "preact";
import { Section, Item } from "@/configurations/types";
import { Component as Cell } from "./configuration-cell";
import { qualified } from "@/configurations";

type Callback = (item: Item, value: any) => void;

export type Props = {
  onUpdated: Callback;
  section: Section;
  itemValueMap: { [p: string]: any };
};

const handleItemUpdated = (onUpdated: Callback) => (value: any, item: Item) => {
  onUpdated(item, value);
};

const makeItem = (key: string, item: Item, value: any, onUpdated: Callback) => {
  return <Cell key={key} item={item} value={value} onUpdated={handleItemUpdated(onUpdated)} />;
};

export const Component: FunctionalComponent<Props> = ({ onUpdated, section, itemValueMap }) => {
  const items = section.items;
  return (
    <div class="configuration-section" data-testid="configuration-section">
      <header class="configuration-section__header" data-testid="configuration-section-header">
        {section.description}
      </header>
      <section class="configuration-section__cell-container">
        {Object.entries(items).map(([key, value]) =>
          makeItem(key, value, itemValueMap[qualified(value.key)], onUpdated)
        )}
      </section>
    </div>
  );
};
