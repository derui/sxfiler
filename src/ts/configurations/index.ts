import { SortType, SortTypeMap } from "@/generated/types_pb";
import { Category, Item, Section, ItemKey, SectionKey, CategoryKey } from "./types";
import { createCategory, createSection, createItem } from "./creators";
import * as ItemCreators from "./item-creators";

const itemKeyFrom = <T>(section: SectionKey, key: string): ItemKey<T> => [section[0], section[1], key];
const categoryKeyOf = (key: ItemKey<any>): CategoryKey => [key[0]];
const sectionKeyOf = (key: ItemKey<any>): SectionKey => [key[0], key[1]];
const sectionKeyFrom = (category: CategoryKey, key: string): SectionKey => [category[0], key];
export const qualified = (key: readonly string[]): string => key.join(".");

export const findItemBy = (key: ItemKey<any>, category: Category): Item | undefined => {
  const categoryKey = categoryKeyOf(key);
  const sectionKey = sectionKeyOf(key);
  if (qualified(category.key) !== qualified(categoryKey)) {
    return;
  }
  const section = category.sections.find((v) => v.key === sectionKey);
  if (!section) {
    return;
  }

  return section.items.find((v) => qualified(v.key) === qualified(key));
};

export const findSectionBy = (key: SectionKey, category: Category): Section | undefined => {
  return category.sections.find((v) => qualified(v.key) === qualified(key));
};

export const findValueBy = (configuration: { [p: string]: any }, item: Item): any => {
  const value = configuration[qualified(item.key)];
  if (value === undefined || value === null) {
    return;
  }

  return value;
};

export const categories = {
  general: ["general"],
} as const;

export const sections = {
  behaviors: sectionKeyFrom(categories.general, "behaviors"),
  theme: sectionKeyFrom(categories.general, "theme"),
} as const;

export const itemKeys = {
  general: {
    behaviors: {
      confirmationWhenDelete: itemKeyFrom<string>(sections.behaviors, "confirmation_when_delete"),
      maxNumberOfHistories: itemKeyFrom<number>(sections.behaviors, "max_number_of_histories"),
      defaultSortOrder: itemKeyFrom<SortTypeMap>(sections.behaviors, "default_sort_order"),
    },
    theme: {
      currentTheme: itemKeyFrom<string>(sections.theme, "current_theme"),
    },
  },
} as const;

export namespace Configurations {
  const general = createCategory({
    key: categories.general,
    displayName: "General",
    description: "General configurations",
    sections: [
      createSection({
        key: sections.behaviors,
        displayName: "Behaviors",
        description: "Behaviors",
        items: [
          createItem({
            key: itemKeys.general.behaviors.confirmationWhenDelete,
            displayName: "Confirmation when delete",
            description: "Need user confirmation when delete something",
            type: ItemCreators.createBoolean(true),
          }),
          createItem({
            key: itemKeys.general.behaviors.maxNumberOfHistories,
            displayName: "Maximum number of histories",
            description:
              "History is the history of moved directory in a pane. This setting defines maximum number of histories in a pane",
            type: ItemCreators.createNumber(100),
          }),
          createItem({
            key: itemKeys.general.behaviors.defaultSortOrder,
            displayName: "Default sort order of Pane",
            description: "This setting defines sort order of filer when open this application",
            type: ItemCreators.createSelectOne(
              [
                {
                  display: "Date",
                  toString() {
                    return `${SortType.DATE}`;
                  },
                  value: SortType.DATE,
                },
                {
                  display: "Name",
                  toString() {
                    return `${SortType.NAME}`;
                  },
                  value: SortType.NAME,
                },
                {
                  display: "Size",
                  toString() {
                    return `${SortType.SIZE}`;
                  },
                  value: SortType.SIZE,
                },
              ],
              `${SortType.NAME}`
            ),
          }),
        ],
      }),
      createSection({
        key: sections.theme,
        displayName: "Theme",
        description: "Theme",
        items: [
          createItem({
            key: itemKeys.general.theme.currentTheme,
            displayName: "Current theme",
            description: "Select current theme",
            type: ItemCreators.createText("default"),
          }),
        ],
      }),
    ],
  });

  export const definitions = {
    general,
  } as const;
}
