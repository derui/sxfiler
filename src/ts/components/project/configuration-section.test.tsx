import { h } from "preact";
import { render as renderAsText } from "preact-render-to-string";
import { Component } from "./configuration-section";
import { render, cleanup, fireEvent, act } from "@testing-library/preact";
import { createSection, createItem } from "@/configurations/creators";
import { createText, createNumber } from "@/configurations/item-creators";

describe("Project", () => {
  describe("Configuration section", () => {
    afterEach(() => cleanup());

    test("show section that do not have any items", () => {
      const section = createSection({
        key: ["cat", "section"],
        displayName: "test",
        description: "description",
        items: [],
      });
      const tree = renderAsText(<Component section={section} itemValueMap={{}} onUpdated={() => {}} />);

      expect(tree).toMatchSnapshot();
    });

    test("show section that have some items", () => {
      const section = createSection({
        key: ["cat", "section"],
        displayName: "test",
        description: "description",
        items: [
          createItem({
            key: ["cat", "section", "item1"],
            displayName: "item1",
            description: "description1",
            type: createText(""),
          }),
          createItem({
            key: ["cat", "section", "item2"],
            displayName: "item2",
            description: "description2",
            type: createNumber(10),
          }),
        ],
      });
      const tree = renderAsText(<Component section={section} itemValueMap={{}} onUpdated={() => {}} />);

      expect(tree).toMatchSnapshot();
    });

    test("callback function if an item updated", async () => {
      const section = createSection({
        key: ["cat", "section"],
        displayName: "test",
        description: "description",
        items: [
          createItem({
            key: ["cat", "section", "item1"],
            displayName: "item1",
            description: "description1",
            type: createText(),
          }),
          createItem({
            key: ["cat", "section", "item2"],
            displayName: "item2",
            description: "description2",
            type: createNumber(),
          }),
        ],
      });
      const callback = jest.fn();
      const query = render(<Component section={section} itemValueMap={{}} onUpdated={callback} />);

      const item1 = await query.findByTestId("configuration-text-cell-input");
      act(() => {
        fireEvent.input(item1, { target: { value: "foobar" } });
      });

      expect(callback).toBeCalledWith(section.items[0], "foobar");
    });
  });
});
