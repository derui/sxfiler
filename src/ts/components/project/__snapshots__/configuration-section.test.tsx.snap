// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Configuration section show section that do not have any items 1`] = `"<div class=\\"configuration-section\\" data-testid=\\"configuration-section\\"><header class=\\"configuration-section__header\\" data-testid=\\"configuration-section-header\\">description</header><section class=\\"configuration-section__cell-container\\"></section></div>"`;

exports[`Project Configuration section show section that have some items 1`] = `"<div class=\\"configuration-section\\" data-testid=\\"configuration-section\\"><header class=\\"configuration-section__header\\" data-testid=\\"configuration-section-header\\">description</header><section class=\\"configuration-section__cell-container\\"><div class=\\"configuration-cell__root\\" data-testid=\\"configuration-text-cell-root\\"><label class=\\"configuration-cell__label\\">item1</label><input class=\\"configuration-cell__text-input\\" data-testid=\\"configuration-text-cell-input\\" type=\\"text\\" value /><p class=\\"configuration-cell__description\\">description1</p></div><div class=\\"configuration-cell__root\\" data-testid=\\"configuration-number-cell-root\\"><label class=\\"configuration-cell__label\\">item2</label><input class=\\"configuration-cell__number-input\\" data-testid=\\"configuration-number-cell-input\\" type=\\"number\\" value=\\"10\\" /><p class=\\"configuration-cell__description\\">description2</p></div></section></div>"`;
